### DISCUSSION TITLE
manipulation of customer invoice counter

### DISCUSSION THREAD

**Discussion by BeckeBauer** [kimai/kimai#3427] [Questions & Answers]
Hi everybody, I am looking for a way to manipulate the invoice counter for a customer (in the database). Reason is that I use the same invoice number with a "C" at the end when I cancel an invoice (e.g. 2022-XXX-01C) instead of the invoice number created by Kimai (which would be 2022-XXX-02 in the example). When I create another invoice for this customer, the individual counter is increased by one (which would be 2022-XXX-03 in the example) so that I have a gap. I might remember about it and manually change the invoice number in the database but forget about it after the next invoice for this customer at the latest. 
I assume that there is a place in the database where the individual counter for each customer is stored but I was not able to find it. Any help to point me in the right direction is appreciated.

**Comment by kevinpapst** [comment]
There is no counter field. The counter is based on the invoice rows in the database.
What exactly are you trying to achieve, I don't get it. 
You know that you can increase the counter, but that is not possible on a "per customer" base. 
Canceled invoices and invoice numbers are indeed a problem, which I do not know how to properly fix.
I tend to delete canceled invoices, so the numbering does not get out of sync.

Any ideas how this could be handled better? 
Allow duplicate invoice numbers? 
Allowing to manually override the generated invoice numbers?

**Comment by BeckeBauer** [comment]
Thank you for your quick feedback. I understand that the "customer-based" counter "counts" the rows in the invoices table that exists for a customer_ID in the calendar year (I assumed there were something like "AUTO_INCREMENT" which I could tinker with). So, no chance to modify the counter without deleting an invoice row - correct?

Regarding your question:  I think that deletion is only an option if the invoice was not yet sent to the customer. Otherwise, a cancellation invoice is required (as far as I understand the tax requirements in Germany). I do not know how other companies handle it but I believe it were correct to have an own number for a cancellation invoice (so that numbering were still in sync and nothing needs to be changed in Kimai). I just prefer to use the same number of the invoice I want to cancel and add a "C" (which makes it quite easy to understand why there is a negative invoice amount).

The easiest way would be to allow overriding the generated invoice number in Kimai directly (if Kimai would somehow "understand" it when creating the next invoice number). As said before, it should not be necessary to have duplicate invoice number which could also cause trouble if wrongly applied. But maybe, the user could get the choice to use the same invoice number with an addition - like "C" - to be defined in the general settings only in case that an invoice is cancelled (and as "add-on" the cancellation invoice is automatically created on basis of the cancelled invoice with negative amounts). And when the defined suffix is found in the table, the invoice counter does not "count" this row when providing the next invoice number...

**Reply by kevinpapst** [reply]
> invoice was not yet sent to the customer

Hm, naaaaah. Everyone creates wrong invoices here and there. Just send the fixed invoice, stating that the last one was wrong. Your customer would deduct a tax fraud, if he would use both. But that is just my personal view and workflow... 😁 

The invoice counter is only a database field, which has just one rule: it needs to be unique. 
We should probably have a look at the next Kimai version. Invoices will receive a lot of improvements, e.g.:
<img width="1290" alt="Bildschirmfoto 2022-07-19 um 20 01 51" src="https://user-images.githubusercontent.com/533162/179818352-69d702e3-3f38-4368-b189-22bdd6355a0b.png">

Customer receive default invoice templates, you can change the invoice date and I could add a field for the invoice number as well. The generated invoice number would be added to the customer specific form and used by default - and you can change it manually for enhanced use-cases.

**Comment by BeckeBauer** [comment]
A field showing the invoice number to be used which can also be manually changed would be quite handy (although probably not needed in the majority of cases)

If it is still calculated on basis of the number of rows in the calendar year and the number of rows of the customer in the calendar year, the 'problem' remained. I don't know if number creation could be realised differently and how much effort such changes would require.

**Reply by kevinpapst** [reply]
> The easiest way would be to allow overriding the generated invoice number in Kimai directly

followed by a 

> the 'problem' remained

I say I could offer a field to adjust the invoice number (what you requested) and you answer it would not solve the issue. I am not sure that I understand what you actually want.

**Comment by BeckeBauer** [comment]
sorry for being unclear - right now, 
Right now, I change the generated invoice number in the database 
Example: 
step 1: I create an invoice for Customer A, Kimai assigns the invoice number 2022-XXX-01 (=first invoice for Customer A)
step 2: One month later, I have to cancel the invoice; to do so, I create another invoice with negative amounts, Kimai assigns the invoice number 2022-XXX-02
step 3; in the database, I change this invoice number to 2022-XXX-01C
step 4: the next time I create an invoice for Customer A, Kimai assigns 2022-XXX-03 as invoice number (not recognising that there is no 2022-XXX-02 anymore)
step 5: I change the invoice number in the database to 2022-XXX-02 (what I usually forget)

So, I thought when you said I could change the invoice number in Kimai directly, it will save step 3 but Kimai would still not be able to adjust the counter. It means step 3 and step 4 still remain (this is how I understood your explanations about how the "Kimai counter" for the individual customer works) - sorry if I misunderstood you here and step 4 and step 5 are also gone with your changes ...

**Reply by kevinpapst** [reply]
I really wonder why people are so obsessed with their invoice numbers. It is not only you, there are many people that are afraid of possible consequences if they do not stick to their "one idea of a correct invoice number".

If there is no legal reason to do it your way, why do you even care how it looks? 
Just let Kimai create a new invoice number for the correction invoice and keep on counting up.

If it is just to get a way to connection between the wrong one and the new one, what do you think about a free text field for new invoices? Like in the screenshot above where you can choose a date, I could add a free text field that allows to add random text to a new invoice. And you would enter "Correction of invoice 2022-XXX-01".

**Reply by BeckeBauer** [reply]
> I really wonder why people are so obsessed with their invoice numbers. It is not only you, there are many people that are afraid of possible consequences if they do not stick to their "one idea of a correct invoice number".

😂 you are right - I think I am afraid of the German Finanzamt and other people might as well. 

> If there is no legal reason to do it your way, why do you even care how it looks?
Just let Kimai create a new invoice number for the correction invoice and keep on counting up.

was told to do it this way by a tax advisor and follow it like a lemming

> If it is just to get a way to connection between the wrong one and the new one, what do you think about a free text field for new invoices? Like in the screenshot above where you can choose a date, I could add a free text field that allows to add random text to a new invoice. And you would enter "Correction of invoice 2022-XXX-01".

like this idea very much, maybe in combination with your former suggestion to manually change the generated invoice number. So, I could add an "C" to the end of the invoice number created by Kimai and enter the free text. Ideally, this free text has a variable that can be shown in the invoice as well!?
